### PR TITLE
Made fields in AbstractRegistry private and added extra methods

### PR DESCRIPTION
**Pull request by Hilbrand**
To get things going I've put this part of the improvements in this pr. This is related to improvements mentioned in #6427. The idea is with this change the next step can be done to improve performance of the AbstractRegistry class and next improve inefficient usages of getAll() and stream(). I've made some changes with the best intentions, but it may be in the context of the architecture it should to be implemented differently.

Changes:
- Fields are made private to protect from implementation detail dependencies.
- Added extra methods to iterate instead of getting fields directly.
- The iteration methods hide implementation and thus can later be implemented effectively when the internal data structure would change.
- The idea is also to get ride of the inefficient getAll(), and the forEach methods can help with that.
- Removed removeManagedProvider it's a duplicate of unsetManagedProvider.

### CODE REVIEW

File: bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/common/registry/AbstractRegistry.java [hunk 1]
```diff
@@ -215,32 +219,34 @@ public E get(K key) {
         return null;
     }
 
+    public Entry<Provider<E>, E> getBoth(K key) {
```
Comment (afuechsel) [inline bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/common/registry/AbstractRegistry.java]: `getBoth` doesn't give good context for what it could be used. Besides, I am missing javadoc for this public method. Maybe `getValueWithProvider` would be a better name?

File: bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/internal/ThingRegistryImpl.java [hunk 2]
```diff
@@ -248,7 +248,13 @@ public Thing createThingOfType(ThingTypeUID thingTypeUID, ThingUID thingUID, Thi
         for (ThingHandlerFactory thingHandlerFactory : thingHandlerFactories) {
             if (thingHandlerFactory.supportsThingType(thingTypeUID)) {
                 Thing thing = thingHandlerFactory.createThing(thingTypeUID, configuration, thingUID, bridgeUID);
-                thing.setLabel(label);
+                if (thing == null) {
```
Comment (afuechsel) [inline bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/internal/ThingRegistryImpl.java]: This change is not related to the PR. Is this fixing something else that you just witnessed? Would be better to create another PR just for fixing this issue.

Comment (Hilbrand) [reply bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/internal/ThingRegistryImpl.java]: Moved to #6433

**Comment by afuechsel** [comment]
You should remove the change you have moved to #6433.

File: bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/common/registry/AbstractRegistry.java [hunk 3]
```diff
@@ -305,12 +333,49 @@ protected void addProvider(Provider<E> provider) {
         return null;
     }
 
+    /**
+     * This method traverses over all elements of a provider in the registry and calls the consumer with each element.
+     *
+     * @param provider provider to traverse elements of
+     * @param consumer function to call with element
+     */
+    public void forEach(Provider<E> provider, Consumer<E> consumer) {
```
Comment (wborn) [inline bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/common/registry/AbstractRegistry.java]: Shouldn't all the new methods have a `protected` modifier instead of a `public` modifier?

Comment (maggu2810) [reply bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/common/registry/AbstractRegistry.java]: Makes sense

**Comment by kaikreuzer** [comment]
@Hilbrand I fear you have to rebase your branch now that @maggu2810's PR has been merged.

**Comment by afuechsel** [comment]
Lgtm
